### PR TITLE
New Examples for Posix Threads + Memory Lint Errors

### PR DESCRIPTION
**Pull request by msaxena2**
WIP. Not yet ready.

### CODE REVIEW

**Comment by msaxena2** [comment]
@dwightguth @grosu Please review.

File: examples/simple/Error_Codes.csv [hunk 1]
```diff
@@ -95,6 +95,14 @@ UB-STDLIB1,"Called realloc on memory not allocated by malloc, or already freed."
 UB-STRING1,Null pointer passed to strcpy.,"7.1.4:1, J.2:1 item 108",Undefined Behavior
 UB-STDLIB2,"Called free on memory not allocated by malloc, or already freed.","7.22.3.3:2, J.2:1 item 179",Undefined Behavior
 UB-TTE1,program requires the formation of a composite type from a variable length array type whose size is specified by an expression that is not evaluated.,"6.2.7:3, J.2:1 item 16",Undefined Behavior
+UB-PTHREAD1,Initializing an already initialiized mutex,POSIX 2008 sec. pthread_mutex_init,Undefined Behavior
```
Comment (dwightguth) [inline examples/simple/Error_Codes.csv]: I'd like to see an extra column with like an X or something in it for those errors that are only caught with RV-Match.

File: examples/simple/lint/L-EXC3-bad.c [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+void func(int number)
+{
+    printf("func called with number: %i\n", number);
+}
+
+int main()
+{
+    void (*func_ptr)(int) = &func;
+    if(func_ptr != ((void *) NULL))
```
Comment (dwightguth) [inline examples/simple/lint/L-EXC3-bad.c]: I think just NULL should be fine because NULL is usually already of type void *

Comment (dwightguth) [inline examples/simple/lint/L-EXC3-bad.c]: also I believe the idea behind this error is it occurs with functions that take no arguments when the user has forgotten to add the () that causes the function to actually be invoked. So I think the good example here should be that we are actually calling func().

File: examples/simple/posix/UB-PTHREAD5-bad.c [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+#include<stdlib.h>
+
+
+pthread_mutex_t * lock_ptr;
+
+
+int main(void)
+{
+    //Calling pthread_mutex_lock without a call to pthread_mutex_init
+    lock_ptr = malloc(sizeof(pthread_mutex_t));
```
Comment (dwightguth) [inline examples/simple/posix/UB-PTHREAD5-bad.c]: I think to make this more interesting what we should do is to initialize the mutex with pthread_mutex_init, but then copy it with memcpy and call the method on the copied mutex.

**Comment by dwightguth** [comment]
looks pretty good. there's a couple things I would do to make this a little more interesting, but otherwise it seems solid to me.

**Comment by msaxena2** [comment]
@dwightguth Please re-review.

**Comment by dwightguth** [comment]
looks good. If the tests pass and @grosu is okay with it go ahead and merge this.

File: examples/simple/lint/L-CEIE1-good.c [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+#include <stdlib.h>
+#include <stdio.h>
+
+int main()
+{
+    int * number_ptr = malloc(sizeof(int));
+    *number_ptr = 10;
+    free(number_ptr);
+    printf("No memory Leaks, Execution Successful\n");
```
Comment (grosu) [inline examples/simple/lint/L-CEIE1-good.c]: Be uniform: either capitalize all words or none.  In fact, since the bad example was minimal, you can also make this one minimal (ie, remove assignment and printf).  Or did you have a reason for which you added the additional statements?

File: examples/simple/lint/L-EXC3-bad.c [hunk 5]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) 2015 Runtime Verification, Inc. (RV-Match team). All Rights Reserved.
+
+#include <stdlib.h>
+#include <stdio.h>
+
+void * func()
+{
+    int * retval = malloc(sizeof(int));
+    *retval = 1;
+    return retval;
```
Comment (grosu) [inline examples/simple/lint/L-EXC3-bad.c]: Don't we have a L-CEIE1 error/warning here, that memory is malloced but not freed?

Comment (dwightguth) [inline examples/simple/lint/L-EXC3-bad.c]: No because this function is never actually called.

Comment (grosu) [inline examples/simple/lint/L-EXC3-bad.c]: I would either remove the malloc statement, or otherwise add a comment in the code saying that the function is not called, so no memory leaks, both in the bad and in the good example. That's because some static analysis tools may report an error here, and we do not want them to think that we "missed" the error.  Something as simple as:

// No memory leak reported, because the function is never invoked.

Comment (dwightguth) [inline examples/simple/lint/L-EXC3-bad.c]: yeah I'd be inclined to just make the function return a pointer to a statically allocated object, eg:
```
int x;
void *func() { return &x; }
```

Comment (dwightguth) [inline examples/simple/lint/L-EXC3-bad.c]: and then do that for both bad and good

File: examples/simple/lint/L-STDLIBE2-good.c [hunk 6]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2015 Runtime Verification, Inc. (RV-Match team). All Rights Reserved.
+
+#include <stdlib.h>
+#include <stdio.h>
+
+int main()
+{
+    size_t array_size = 20;
```
Comment (grosu) [inline examples/simple/lint/L-STDLIBE2-good.c]: Does this need to be size_t instead of int?  One would expect the -good programs to only fix the -bad ones minimally.

Comment (dwightguth) [inline examples/simple/lint/L-STDLIBE2-good.c]: good catch

File: examples/simple/lint/L-STDLIBE3-good.c [hunk 7]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2015 Runtime Verification, Inc. (RV-Match team). All Rights Reserved.
+
+#include <stdlib.h>
+#include <stdio.h>
+
+int main()
+{
+    size_t array_size = 20;
```
Comment (grosu) [inline examples/simple/lint/L-STDLIBE3-good.c]: same here

File: examples/simple/posix/UB-PTHREAD2-bad.output [hunk 8]
```diff
@@ -0,0 +1,12 @@
+Error: UB-PTHREAD2
+Description: Destroying a locked mutex.
+Locked here:
+  at pthread_mutex_lock(UB-PTHREAD2-bad.c:15)
+  at locker_thread(UB-PTHREAD2-bad.c:15)
+  at <file-scope>(<unknown>)
+
```
Comment (grosu) [inline examples/simple/posix/UB-PTHREAD2-bad.output]: Is this white line necessary?  I initially thought that there are two error messages.  Others may be confused, too.

File: examples/simple/posix/UB-PTHREAD4-bad.output [hunk 9]
```diff
@@ -0,0 +1,12 @@
+Error: UB-PTHREAD4
+Description: Attempting to lock a PTHREAD_MUTEX_DEFAULT mutex that is already locked by the current thread.
+Locked here:
+  at pthread_mutex_lock(UB-PTHREAD4-bad.c:15)
+  at locker_thread(UB-PTHREAD4-bad.c:15)
+  at <file-scope>(<unknown>)
+
```
Comment (grosu) [inline examples/simple/posix/UB-PTHREAD4-bad.output]: same here: is the whiteline necessary, or follows some more general convention?  if not, i would remove it.

File: examples/simple/posix/UB-PTHREAD5-good.c [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+#include <pthread.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+pthread_mutex_t lock;
+
+int main(void)
+{
+
```
Comment (grosu) [inline examples/simple/posix/UB-PTHREAD5-good.c]: any reason for this whiteline?  you typically didn;t have whitelines at the beginning of functions.  also, not clear to me when you put one whiteline and when you put more.  no big deal, but in general it is a good idea to follow consistent formatting and rules, because that shows seriousness and professionalism.

File: examples/simple/posix/UB-PTHREAD2-bad.output (hunk 8 above)
Comment (dwightguth) [inline examples/simple/posix/UB-PTHREAD2-bad.output]: @msaxena2 go ahead and remove this line and I will fix the code in the upcoming pull request I am working on that deals with examples.

File: examples/simple/posix/UB-PTHREAD6-bad.output [hunk 11]
```diff
@@ -0,0 +1,7 @@
+Error: UB-PTHREAD6
+Description: Attempting to unlock a PTHREAD_MUTEX_DEFAULT or PTHREAD_MUTEX_NORMAL mutex which is not robust and which has not been locked by the current thread.
```
Comment (grosu) [inline examples/simple/posix/UB-PTHREAD6-bad.output]: This description is a bit hard to follow.  Can you remove `which is not robust and`?  Since it is an undefined behavior, it is sort of understood that it is not robust.  Or am I misunderstanding it?

File: examples/simple/posix/UB-PTHREAD6-bad.c [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+#include<stdlib.h>
+
+
+
+pthread_t thread_id;
+pthread_mutex_t lock;
+
+void * locker_thread(void * ptr) 
+{
+    //unlocking an unlocked thread
```
Comment (grosu) [inline examples/simple/posix/UB-PTHREAD6-bad.c]: I am all for such comments, but we should aim at consistency: either we add such comments to all -bad examples, or to none, and we should either start all of them with a capital letter or none, and we should have a period at the end everywhere or nowhere.  actually I'm not very sure about the first, it kind of makes sense to add such comments selectively, where things are tricky.

File: examples/simple/posix/UB-PTHREAD8-good.c [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+   
+    
+    pthread_join(thread_id, NULL);
+    pthread_mutex_destroy(&lock);
+    
+    printf("Successful Execution\n");
+    
+    return 0;
+}
```
Comment (grosu) [inline examples/simple/posix/UB-PTHREAD8-good.c]: The -good one seems a lot larger and different than the -bad one.  Can they get closer to each other?  I think people will like to see quickly the fix.

File: examples/simple/posix/UB-PTHREAD6-bad.output (hunk 11 above)
Comment (dwightguth) [inline examples/simple/posix/UB-PTHREAD6-bad.output]: @grosu  a robust mutex is a special type of mutex in POSIX 2008. Attempting to unlock a robust mutex does not result in undefined behavior, hence the error message. If you can think of another way of rephrasing this, I am open to suggestions.

**Comment by grosu** [comment]
I have no other comments.  Very nice examples!  In fact, I think having people go through our canonical examples is a good idea, because they both learn C better and appreciate RV-Match more.  I'm going to add a sentence or two about that in the documentation.  Thank you, guys, very good work!

File: examples/simple/posix/UB-PTHREAD6-bad.output (hunk 11 above)
Comment (grosu) [inline examples/simple/posix/UB-PTHREAD6-bad.output]: But the type of the error is Undefined behavior.  If this was a warning and the type potentially undesired behavior, then I would have raised no objections.

Comment (dwightguth) [inline examples/simple/posix/UB-PTHREAD6-bad.output]: It is considered undefined behavior, yes. Specifically it is considered undefined by the POSIX 2008 standard to unlock a mutex you do not own and have locked, unless the mutex is a  robust mutex, in which case it is well defined and returns a specific error code.

Comment (grosu) [inline examples/simple/posix/UB-PTHREAD6-bad.output]: OK then.  What I didn't know was that `robust` is a technical term in this context.  Hopefully others will not get confused.

**Comment by msaxena2** [comment]
@dwightguth @grosu Made some more changes. Please re-review.

File: examples/simple/lint/L-CCVE2-bad.output [hunk 14]
```diff
@@ -0,0 +1,5 @@
+Warning: L-CCVE2
+Description: conversion to unsigned integer outside the range that can be represented.
```
Comment (grosu) [inline examples/simple/lint/L-CCVE2-bad.output]: The description is still not capitalized.

**Comment by dwightguth** [comment]
looks good. seems like you addressed all the comments. You tested all the examples and they seem to work as intended? If so, and if @grosu is okay, I am fine with this being merged.

File: examples/simple/lint/L-CEIE1-bad.c [hunk 15]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) 2015 Runtime Verification, Inc. (RV-Match team). All Rights Reserved.
+
+#include <stdlib.h>
+#include <stdio.h>
+
+int main()
```
Comment (grosu) [inline examples/simple/lint/L-CEIE1-bad.c]: You had two lines before the main function until now, and now you have one line.

@msaxena2 , can you please go through my previous comments and address them one by one, telling whether you took care of it and how, or whether you have not and why?  Otherwise I'm wasting a lot of time re-checking everything that I've already reported.

File: examples/simple/lint/L-CCVE1-bad.output [hunk 16]
```diff
@@ -0,0 +1,5 @@
+Warning: L-CCVE1
+Description: unsigned integer overflow.
```
Comment (dwightguth) [inline examples/simple/lint/L-CCVE1-bad.output]: this and the one @grosu mention should be capitalized. I will make a PR to handle the code change.

File: examples/simple/lint/L-EXC3-bad.c [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+    return &retval; 
+}
+
+int main()
+{
+    void * (*func_ptr)() = &func;
+    if(func_ptr != NULL)
+    {
+       printf("Hello, World!\n"); 
+
```
Comment (grosu) [inline examples/simple/lint/L-EXC3-bad.c]: why a white line here?

File: examples/simple/lint/L-PTHREAD9-bad.output [hunk 18]
```diff
@@ -0,0 +1,8 @@
+Warning: L-PTHREAD9
+Description: Terminating a thread while holding lock on mutexes. Locked here:
+  at pthread_mutex_lock(L-PTHREAD9-bad.c:15)
+  at locker_thread(L-PTHREAD9-bad.c:15)
+  at <file-scope>(<unknown>)
+
```
Comment (grosu) [inline examples/simple/lint/L-PTHREAD9-bad.output]: why whiteline?

File: examples/simple/lint/L-PTHREAD9-good.c [hunk 19]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2015 Runtime Verification, Inc. (RV-Match team). All Rights Reserved.
+
+#include<stdio.h>
+#include<string.h>
+#include<pthread.h>
+#include<stdlib.h>
+
+
```
Comment (grosu) [inline examples/simple/lint/L-PTHREAD9-good.c]: why 3 whitelines?

File: examples/simple/lint/L-STDLIBE2-good.c (hunk 6 above)
Comment (grosu) [inline examples/simple/lint/L-STDLIBE2-good.c]: so, do we change this to int or not?

File: examples/simple/lint/L-STDLIBE2-good.c [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+    int * small_int_array = malloc(array_size * sizeof(int));
+    if(small_int_array != NULL)
+    {
+
```
Comment (grosu) [inline examples/simple/lint/L-STDLIBE2-good.c]: why whiteline?

File: examples/simple/lint/L-STDLIBE3-bad.c [hunk 21]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2015 Runtime Verification, Inc. (RV-Match team). All Rights Reserved.
+
+#include <stdlib.h>
+#include <stdio.h>
+
+int main()
+{
+    size_t array_size = 1024 * 1024 * 1024;
```
Comment (grosu) [inline examples/simple/lint/L-STDLIBE3-bad.c]: int?

File: examples/simple/posix/UB-PTHREAD1-bad.c [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <stdio.h>
+#include <pthread.h>
+#include <stdlib.h>
+
+pthread_mutex_t lock;
+
+int main(void)
+{
+
```
Comment (grosu) [inline examples/simple/posix/UB-PTHREAD1-bad.c]: why whiteline?

File: examples/simple/posix/UB-PTHREAD6-bad.c [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+int main(void)
+{
+    int err = 0;
+    if (pthread_mutex_init(&lock, NULL) != 0)
+    {
+        printf("\n mutex init failed\n");
+        return 1;
+    }
+
+
```
Comment (grosu) [inline examples/simple/posix/UB-PTHREAD6-bad.c]: 3 whitelines?

File: examples/simple/lint/L-CEIE1-good.c (hunk 4 above)
Comment (msaxena2) [inline examples/simple/lint/L-CEIE1-good.c]: I agree the printf was unnecessary; it's been removed.

File: examples/simple/lint/L-EXC3-bad.c (hunk 5 above)
Comment (msaxena2) [inline examples/simple/lint/L-EXC3-bad.c]: Malloc has been replaced by a statically allocated object in both -good and -bad versions.

File: examples/simple/lint/L-STDLIBE2-good.c (hunk 6 above)
Comment (msaxena2) [inline examples/simple/lint/L-STDLIBE2-good.c]: for now, I've kept size_t in both good and bad versions. We can change it if needed.

File: examples/simple/lint/L-CEIE1-bad.c (hunk 15 above)
Comment (msaxena2) [inline examples/simple/lint/L-CEIE1-bad.c]: I've added another whitespace here to adhere to a consistent format, described in one of the later comments.

File: examples/simple/lint/L-PTHREAD9-good.c (hunk 19 above)
Comment (msaxena2) [inline examples/simple/lint/L-PTHREAD9-good.c]: Changed to 2 whitespaces.

File: examples/simple/lint/L-STDLIBE2-good.c (hunk 20 above)
Comment (msaxena2) [inline examples/simple/lint/L-STDLIBE2-good.c]: Removed the whitespace here.

File: examples/simple/posix/UB-PTHREAD1-bad.c (hunk 22 above)
Comment (msaxena2) [inline examples/simple/posix/UB-PTHREAD1-bad.c]: Addressed by removing the whiteline.

File: examples/simple/posix/UB-PTHREAD6-bad.c (hunk 23 above)
Comment (msaxena2) [inline examples/simple/posix/UB-PTHREAD6-bad.c]: Addressed by removing the whiteline.

**Comment by msaxena2** [comment]
@grosu @dwightguth Please re-review. Changes made since previous review:
* I've addressed all the whiteline related comments that had been made in the previous review. The files now have 2 whitelines between the last '#include' and the following line of code, and a single whiteline between different function definitions. Within function blocks, I've removed random occurrences of whitelines. I've done this for all files to ensure uniformity.
* Both good and bad version of STDLIBE2, STDLIBE3 now use size_t instead of int. I'm open to reverting them back to int.

**Comment by grosu** [comment]
lgtm
